craft_providers/base.py

```diff
@@ -49,10 +49,19 @@ class Base(ABC):
         extend this tag, not overwrite it, e.g.: compatibilty_tag =
         f"{appname}-{Base.compatibility_tag}.{apprevision}" to ensure base
         compatibility levels are maintained.
+
+    :param environment: Command environment.
     """
 
     compatibility_tag: str = "base-v0"
 
+    def __init__(
+        self,
+        *,
+        environment: Dict[str, Optional[str]],
```

Comment:
It's strange that this parameter is mandatory here, and optional in the inherited one. 

What would be the behaviours of other classes inheriting this? Because maybe we could make it optional everywhere, and *here* get the `default_command_environment()` if none was passed.

Reply:
I considered that, but default_command_environment() in theory is specific BuilddBase.  If it were say a Windows instance (e.g. `WindowsBase`), that implementation would define its own default_command_environment.  The common denominator is that the base `environment` should always be defined for a base, but each base may carry its own default (allowing it to be optional in the subclass).

Reply:
from an API point of view I would instead do

```python
@abc.abstractmethod
def get_environment(self) -> Dict[str, Optional[str]]
    """Return the command environment."""
```

to require its implementation in the classes that inherit it

Reply:
Fair point, done!